NetSSL_OpenSSL/src/HTTPSClientSession.cpp

```diff
@@ -117,6 +119,22 @@ void HTTPSClientSession::abort()
 }
 
 
+StreamSocket HTTPSClientSession::detachSocket()
+{
+	Poco::Net::SecureStreamSocket *oldSocket = dynamic_cast<Poco::Net::SecureStreamSocket *>(&socket());
+	if (oldSocket)
+	{
+		Poco::Net::SecureStreamSocket newSocket(oldSocket->context());
+		attachSocket(newSocket);
+		return *oldSocket;
+	}
+	else
+	{
+		throw InvalidArgumentException("Socket is not SecureStreamSocket");
```

Comment:
I would not use `InvalidArgumentException` here, because `detachSocket()` has no arguments. This is looks rather like a `LogicException` to me (you found the object in a state that you did not expect).

For a production code I would say that the case `socket()` not being a `SecureStreamSocket` is not really an exception -- you just create the new socket as a `StreamSocket` and don't set context (as the original code does). But since it is a hack it is not important.

Reply:
https://pocoproject.org/docs/Poco.Net.HTTPSession.html#24153
```c++
StreamSocket & socket();
```
It returns a **reference** to the socket. Then you get a pointer to the same object cast to a `SecureStreamSocket` in `oldSocket`. Then you call `detachSocket()` -- this destroys the object pointed by `oldSocket`. Then you use `oldSocket` in the return statement, but it does not point to a valid object anymore => undefined behavior!!